prototype/distributor/src/ServerHandler.java

```diff
@@ -55,188 +78,168 @@ public boolean onHandshake(Control.Handshake handshake_packet) {
+        Control.SessionRequest.Builder srbuilder = Control.SessionRequest.newBuilder();
+        srbuilder.setName(selfName);
+        Control.SessionRequest newRequest = srbuilder.build();
+        sendProto(newRequest.toByteString(), General.FvPacketType.SESSION_REQUEST, UsersData.getUser(sr.getName()));
+    }
 
-            // set outer packet
-            replyOuter = General.FvPacket.newBuilder()
-                             .setType(General.FvPacketType.SERVER_MESSAGE)
-                             .setInnerPacket(ByteString.copyFrom(reply.toByteArray()))
-                             .build();
+    public void onSessionResponse(Control.SessionResponse sr) {
+        Control.ServerMessage reply;
 
-            sendProto(replyOuter);
+        System.out.println("Session Response packet");
 
+        //Connection refused by host
+        if (!sr.getResponse()){
+            reply = Control.ServerMessage.newBuilder()
+                    .setType(Control.ServerMessageType.SESSION_REJECTED)
+                    .setSuccess(false)
+                    .build();
+            sendProto(ByteString.copyFrom(reply.toByteArray()), General.FvPacketType.SERVER_MESSAGE);
             return;
         }
+
         for (SessionData session : UsersData.getSessionList()) {
             // if user is already hosting a session
-            if (session.getHostUser().equals(sr.getName())) {
-                // add current user to requested user's session
-                session.addClientUser(selfName);
+            if (session.getHostUser().equals(selfName)) {
+                // add requesting user to current user's session
+                session.addClientUser(sr.getName());
 
                 // user is available: send session starting to requester
                 reply = Control.ServerMessage.newBuilder()
-                            .setType(Control.ServerMessageType.SESSION_STARTING)
-                            .setSuccess(true)
-                            .build();
-
-                // set outer packet
-                replyOuter = General.FvPacket.newBuilder()
-                                 .setType(General.FvPacketType.SERVER_MESSAGE)
-                                 .setInnerPacket(ByteString.copyFrom(reply.toByteArray()))
-                                 .build();
-
-                sendProto(replyOuter);
-
-                // send new user video params
-                replyOuter =
-                    General.FvPacket.newBuilder()
-                        .setType(General.FvPacketType.VIDEO_PARAMS)
-                        .setInnerPacket(ByteString.copyFrom(session.getVideoParams().toByteArray()))
+                        .setType(Control.ServerMessageType.SESSION_STARTING)
+                        .setSuccess(true)
                         .build();
+                sendProto(ByteString.copyFrom(reply.toByteArray()), General.FvPacketType.SERVER_MESSAGE, UsersData.getUser(sr.getName()));
```

Comment:
In the case of a "race condition" here, do some error checking to ensure that the requester still exists.

Reply:
same as L212, 213

Reply:
What is the race condition here? I'm not sure I understand. I added a check at the beginning of the method to ensure that the "client" still exists by the time the server receives the host's response packet.

Resolved (maybe).

Reply:
that check should resolve it